sdks/python/apache_beam/io/vcfio.py

```diff
@@ -0,0 +1,378 @@
+        `phaseset` is present. If a genotype is not called (that is, a `.` is
+        present in the GT string), -1 is used
+      phaseset (str): If this field is present, this variant call's genotype
+        ordering implies the phase of the bases and is consistent with any other
+        variant calls in the same reference sequence which have the same
+        phaseset value. If the genotype data was phased but no phase set was
+        specified, this field will be set to `*`.
+      info (dict): A map of additional variant call information. The key is
+        specified in the VCF record and the type of the value is specified by
+        the VCF header FORMAT.
+    """
+    self.name = name
+    self.genotype = genotype or []
+    self.phaseset = phaseset
+    self.info = info or {}
+
+  def __eq__(self, other):
+    return (self.name == other.name and
+            self.genotype == other.genotype and
+            self.phaseset == other.phaseset and
+            self.info == other.info)
+
+  def __repr__(self):
+    return ', '.join(
+        [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
+
+
+class _VcfSource(filebasedsource.FileBasedSource):
+  r"""A source for reading VCF files.
+
+  Parses VCF files (version 4) using PyVCF library. If file_pattern specifies
+  multiple files, then the header from each file is used separately to parse
+  the content. However, the output will be a uniform PCollection of
+  :class:`Variant` objects.
+  """
+
+  DEFAULT_VCF_READ_BUFFER_SIZE = 65536  # 64kB
+
+  def __init__(self,
+               file_pattern,
+               min_bundle_size=0,
+               compression_type=CompressionTypes.AUTO,
+               buffer_size=DEFAULT_VCF_READ_BUFFER_SIZE,
+               validate=True):
+    super(_VcfSource, self).__init__(file_pattern, min_bundle_size,
+                                     compression_type=compression_type,
+                                     validate=validate)
+
+    self._coder = coders.StrUtf8Coder() # VCF files are UTF-8 per schema.
+    self._header_lines_per_file = {}
+    self._text_source = TextSource(
+        file_pattern,
+        min_bundle_size,
+        compression_type,
+        strip_trailing_newlines=True,
+        coder=self._coder,
+        buffer_size=buffer_size,
+        validate=validate,
+        skip_header_lines=0,
+        header_matcher_predicate=lambda x: x.startswith('#'))
+
+  def _process_header_lines_matching_predicate(self, file_name, header_lines):
+    self._header_lines_per_file[file_name] = header_lines
+
+  def read_records(self, file_name, range_tracker):
+    def line_generator():
+      header_processed = False
+      for line in self._text_source.read_records(
+          file_name,
+          range_tracker,
+          self._process_header_lines_matching_predicate):
+        if not header_processed and file_name in self._header_lines_per_file:
+          for header in self._header_lines_per_file[file_name]:
```

Comment:
<!--new_thread; commit:7967f9db74c71484c9b94f42bf5c5219cbb6cb85; resolved:0-->
Seems like this will cause header to be returned for every bundle of a file which is prob. not the expected behavior. You should limit it to the first bundle (that will start with offset 0).

Reply:
We need the header in each bundle since it contains important metadata required to read the contents of the file.

Reply:
So to clarify, what do you want PCollection produced by reading VcFSource to contain ? Based on current implementation any given line of the PCollection can be a header line or a data line. A runner might choose to logically split files at arbitrary locations to parallelize reading. I think it will be hard to use the PCollection produced in the current form. A better solution might be to produce a PCollection of records where each record contains the header instead of yielding the header lines when reading.

Reply:
Also, please address this.